xdsl/utils/parse_pipeline.py

```diff
@@ -0,0 +1,82 @@
+import re
+from dataclasses import dataclass
+from typing import Iterator
+from xdsl.utils.lexer import Input, Span
+from enum import Enum, auto
+
+
+class Kind(Enum):
+    EOF = auto()
+
+    IDENT = auto()
+    L_BRACE = auto()
+    R_BRACE = auto()
+    EQUALS = auto()
+    NUMBER = auto()
+    SPACE = auto()
+    STRING_LIT = auto()
+    COMMA = auto()
+
+
+@dataclass
+class Token:
+    span: Span
+    kind: Kind
+
+
+_lexer_rules: list[tuple[re.Pattern[str], Kind]] = [
+    (re.compile(r"[-+]?[0-9]+(\.[0-9]*([eE][-+]?[0-9]+)?)?"), Kind.NUMBER),
+    (re.compile(r"[A-Za-z0-9_-]+"), Kind.IDENT),
+    (re.compile(r'"(\\[nfvtr"\\]|[^\n\f\v\r"\\])*"'), Kind.STRING_LIT),
+    (re.compile(r"\{"), Kind.L_BRACE),
+    (re.compile(r"}"), Kind.R_BRACE),
+    (re.compile(r"="), Kind.EQUALS),
+    (re.compile(r"\s+"), Kind.SPACE),
+    (re.compile(r","), Kind.COMMA),
+]
+"""
+This is a list of lexer rules that should be tried in this specific order to get the next token.
+"""
+
+
+def tokenize_pass(input_str: str) -> Iterator[Token]:
+    """
+    This tokenizes a pass declaration string. Pass syntax is a subset
+    of MLIRs pass pipeline syntax:
+
+    pipeline          ::= pipeline-element (`,` pipeline-element)*
+    pipeline-element  ::= pass-name options?
+    options           ::= `{` options-element ( ` ` options-element)* `}`
+    options-element   ::= key `=` value
```

Comment:
Could we have
```
options-element ::= key ( `=` value )
```
here? It is used in MLIR and useful at times!

Reply:
I was considering that, but I am unsure what the value should be in that case. Boolean?

I would do that in a later PR, as I fear this to be a bit of a discussion point...

Reply:
Oh okay, I don't expect it but better keep this first one as clear as possible sure

Reply:
We do have this now :D